rust/gui-client/src-tauri/src/gui.rs

```diff
@@ -396,40 +420,53 @@ async fn smoke_test(ctlr_tx: CtlrTx) -> Result<()> {
-                        .context("Incoming deep link was not valid UTF-8")?,
-                )
-                .context("Impossible: can't wrap String into SecretString")?;
-                // Ignore errors from this, it would only happen if the app is shutting down, otherwise we would wait
-                ctlr_tx
-                    .send(ControllerRequest::SchemeRequest(url))
-                    .await
-                    .ok();
-            }
-            Ok(None) => {
-                tracing::debug!("Accepted deep-link but read 0 bytes, trying again ...");
-            }
-            Err(error) => {
-                tracing::warn!("Failed to accept deep link: {error:#}")
-            }
-        }
-        // We re-create the named pipe server every time we get a link, because of an oddity in the Windows API.
-        server = deep_link::Server::new().await?;
-    }
-}
-
 fn handle_system_tray_event(app: &tauri::AppHandle, event: system_tray::Event) -> Result<()> {
     app.try_state::<Managed>()
         .context("can't get Managed struct from Tauri")?
         .ctlr_tx
         .blocking_send(ControllerRequest::SystemTrayMenu(event))?;
     Ok(())
 }
+
+/// Create a new instance of the GUI IPC server.
+///
+/// Every time Firezone gets launched, we attempt to connect to this server.
+/// If we can successfully connect and handshake a message, then we know that there is a functioning instance of Firezone already running.
+///
+/// If we hit an IO errors during the connection, we assume that there isn't already an instance of Firezone and we consider us to be the first instance.
+///
+/// There is a third, somewhat gnarly case:
+///
+/// An instance of Firezone may already be running but it is not responding.
+/// Launching another instance on top of it would likely create more problems that it solves, so we also need to fail for that case.
+async fn create_gui_ipc_server() -> Result<ipc::Server> {
+    let (mut read, mut write) = match ipc::connect::<ServerMsg, ClientMsg>(
+        SocketId::Gui,
+        ipc::ConnectOptions { num_attempts: 1 },
+    )
+    .await
+    {
+        Ok(pair) => pair,
+        Err(e) => {
+            // If we can't connect to the socket, we must be the first instance.
+            tracing::debug!(
+                "We appear to be the first instance of the GUI client; connecting to socket yielded: {e:#}"
+            );
+
+            return ipc::Server::new(SocketId::Gui).context("Failed to create GUI IPC socket");
+        }
+    };
+
+    write.send(&ClientMsg::NewInstance).await?;
+    let response = read
+        .next()
+        .await
+        .context("No response received")?
+        .context("Failed to receive response")?;
+
+    anyhow::ensure!(response == ServerMsg::Ack);
+
+    // If we managed to send the IPC message then another instance of Firezone is already running.
+
+    bail!("Successfully handshaked with existing instance of Firezone GUI")
+}
```

Comment:
@jamilbk Happy to chat through this to see if you have any better ideas but this is the most robust I could come up with. The tricky bit here is that we need to make sure that we launch a new IPC server, even if a previous run has left behind a socket. Then connection to the socket will fail and we need to actually overwrite it.

However, if we successfully connect but somehow the instance is not responding to our handshake, we should not launch another server.